File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();

        int left = 0;
        int maxL = 0;


        for(int i = 0; i < s.length(); i++){
            if(map.containsKey(s.charAt(i)) && map.get(s.charAt(i)) >= left){
                left = map.get(s.charAt(i)) + 1;
            } else {
                maxL = Math.max(maxL, i - left + 1);
            }
            map.put(s.charAt(i), i);
        }
        return maxL;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        map = {}

        maxL = 0
        left = 0

        for i, char in enumerate(list(s)):
            if char in map and map.get(char) >= left:
                left = map.get(char) + 1
            else:
                maxL = max(maxL, i - left + 1)
            map[char] = i

        return maxL



